### PR TITLE
Adhere to V.1.0.0 new global scoping rules.

### LINKED ISSUE #6: UndefVarError: m_x_t_min not defined
I ran into this error while running the **1_state_estimation_forward_only demo** as a `.jl` script inside the Atom Juno IDE. I'm running Ubuntu 18.04.

This is expected behavior since Julia v. 1.0.0. as noted here: JuliaLang/julia#28523

In order to access global variables inside blocks, they must be specified with the `global` keyword inside the block. 

This snippet reproduces the error:
```
i = 10
while i > 0
    println(i -= 1)
end 

UndefVarError: i not defined
```

This is the fix:
```
i = 10
while i > 0
    global i
    println(i -= 1)
end
```
Note that this error does not occur with IJulia notebooks. Apparently, this behavior is defined differently for interactive evaluation contexts: JuliaLang/julia#28789.

### PR DESCRIPTION
**Pull request by mroavi**
Resolves #6 by adding  `global` specifiers to global variables inside blocks where they are accessed.

### CODE REVIEW

**Comment by ThijsvdLaar** [comment]
Thanks for the PR, I still have an issue though, which is also a point of discussion.

In think the use of global variables should be reduced to a minimum, and preferrably they should not be used at all. I think it's better to use syntax that avoids writing to the outer scope, e.g.
```
x_0 = 1
for t=1:T
  if t==1
    x_t_min = x_0
  else
    x_t_min = x[t-1]
  end
  ...
end
```
or a shorthand of this.

**Comment by mroavi** [comment]
Hi Thijs. I read the documentation on [Scope of Variables](https://docs.julialang.org/en/v1/manual/variables-and-scoping/#scope-of-variables-1). According to me the `x_0` in the code you posted is still a global variable inside the main module. However, I did verify that your code does not throw an error. The reason is that writing to global variables inside local scopes is not permitted but reading them is ok:

> Avoiding globals
> Avoiding changing the value of global variables is considered by many to be a programming best-practice. One reason for this is that remotely changing the state of global variables in other modules should be done with care as it makes the local behavior of the program hard to reason about. This is why the scope blocks that introduce local scope require the global keyword to declare the intent to modify a global variable.

The code you posted is another way of fixing the error, but it does not solve the problem of not using global variables.

And I agree with the fact that writing global variables in local scopes is perhaps not the best solution. However, this is just a simple demo that is not meant to be used or incorporated as part of other larger software projects. Let me know what you think.

**Comment by ThijsvdLaar** [comment]
I see; the issue is not with _using_ a global variable in the local scope, but with _writing_ to a global variable from the local scope. Using global variables in local scopes is not a bad thing, and you would even be unable to write practical code without it. However, the `global` keyword effectively allows one to override best practices by allowing one to write to global variables from the local scope. I think that's not something we want to advocate, even if it is just for a demo.

**Comment by mroavi** [comment]
Yes, I agree. I will update the fix and use your approach instead.

**Comment by mroavi** [comment]
I found another way to fix this. The `let` block can be used to save the value of a global variable in a local one.

For example, this fixes the error in demo 1:
```
let m_x_t_min = m_x_0, v_x_t_min = v_x_0
    for t = 1:n_samples
        # Prepare data and prior statistics
        data = Dict(:y_t       => y[t],
                    :m_x_t_min => m_x_t_min,
                    :v_x_t_min => v_x_t_min)

        # Execute algorithm
        marginals = step!(data)

        # Extract posterior statistics
        m_x_t_min = mean(marginals[:x_t])
        v_x_t_min = var(marginals[:x_t])

        # Store to buffer
        m_x[t] = m_x_t_min
        v_x[t] = v_x_t_min
    end
end
```
Here m_x_t_min is local while  m_x_0 is global. 

I like this approach because it avoids changing the control flow of the program and hence there is less risk of introducing a new bug. The let block is explained in the [Scope of Variables](https://docs.julialang.org/en/v1/manual/variables-and-scoping/#scope-of-variables-1) documentation. Any thoughts?

**Comment by MagnusKoudahl** [comment]
I ran into this today as well and spent some time on fixing it before finding this issue. Are we merging a fix soon?

**Comment by bertdv** [comment]
OK @MagnusKoudahl @mroavi thanks for working this out. @ivan-bocharov can you please review the PR and merge if adequate fix? thanks. I think in more general terms, let's talk about how to avoid delays like this after we all get back. I think we need to make some agreements on how to scale FL development to group level :) Also, finalize the docs that @ivan-bocharov wrote on how to develop for FL.

**Comment by ivan-bocharov** [comment]
I have reviewed this PR and have following comments:

First of all, changes requested by @ThijsvdLaar are not reflected in this PR (it was not updated after discussion) and I do not think it can be merged until we reach an agreement on how to handle scoping in demos. It is also worth to note that the problem only occurs in demos that were converted to a script and not in Jupyter notebooks.

Second, I do not think that using `global` keyword is inherently wrong. As mentioned by @mroavi scoping rules for `for`, `while`, `try-catch-finally`, `let` blocks have changed in 0.7/1.0 in order to make static locality prediction possible in all cases. We were writing to a "global" variable in previous versions of the demo as well, the only difference is that now we have to preface this variable with a `global` keyword. I don't agree that it necessarily promotes bad programming practices.

However, if there's strong opposition to using `global` keyword in demos, we essentially have two equivalent options (example from `1_state_estimation_forward_only`):

```julia
m_x_t_min = m_x_0
v_x_t_min = v_x_0
for t = 1:n_samples

    # Extract prior statistics
    m_x_t_min = t==1 ? m_x_0 : m_x[t-1]
    v_x_t_min = t==1 ? v_x_0 : v_x[t-1]

    # Prepare data and prior statistics
    data = Dict(:y_t       => y[t],
                :m_x_t_min => m_x_t_min,
                :v_x_t_min => v_x_t_min)
    
    # Execute algorithm
    marginals = step!(data)

    # Extract posterior statistics and store them to buffer 
    m_x[t] = mean(marginals[:x_t])
    v_x[t] = var(marginals[:x_t])
end
```
and wrapping everything in a `let` block as proposed by @mroavi. The first approach doesn't introduce otherwise unnecessary block but is also a bit verbose in prior statistics extraction. For me, using `global` or ternary operator variants are equally ok, while I like `let` block solution a bit less.

In conclusion, let's decide on how we would like to address scoping in demos. After that @mroavi or @MagnusKoudahl can fix the demos accordingly and I will merge this PR.

**Comment by mroavi** [comment]
I am in favor of the `global` keyword solution since it adds the least noise to the code. The other two solutions are good options as well.

**Comment by ThijsvdLaar** [comment]
Thanks for all your comments, @marcocox , @bertdv and I just had a discussion about this and came to the following conclusion:
The reported problems occur when copying the code from the a notebook to a file that gets executed in the REPL. In this case, high level variables are placed within the global scope and are not writeable. It is however the choice of the person that wants to execute the code in this particular fashion, and not a problem with the underlying code base. Therefore, we decided to not change the code base and make it the responsibility of the `executer` to wrap his code in an encompassing module, which would solve the issue.

**Comment by mroavi** [comment]
Alright. I will close the issue. Thanks, everyone.

**Comment by MagnusKoudahl** [comment]
I encountered the issue working in Atom, unrelated to using a notebook.
If we are going to leave the code base as is, may I suggest adding in a comment/hint/suggestion in the tutorial, readme or another obvious place? When people start using Forneylab they are likely to encounter this and the solution should be obvious from the getgo
